`evaluate_composition` visualizer

nbokulich opened the pull request:
do not merge until after 2017.10 release

jairideout commented:
FYI, I disabled the noisy Coveralls comments on this repo, so coverage status will only be reported using GitHub's status API.

nbokulich commented:
Thanks @gregcaporaso ! I have addressed all of the urgent changes that you recommended above. I have created issues for the "feature request" comments because of the tight deadline on this release, and can take care of these next release (and some fit in with what I already had planned but was holding off for this release).

> You should probably add descriptions of the different metrics in the QZV or in documentation of this visualizer.

I have added a little more description to the docs; I think this may still be a little lacking but plan to describe these a little more in the tutorial. Alternatively, I could describe these metrics in the readme for this plugin, then link to that readme from the tutorials, docs, and other places as appropriate (I think I like that approach best). What do you think @gregcaporaso ?

gregcaporaso commented:
@nbokulich, I think this all sounds good for now.

thermokarst commented on q2_quality_control/_evaluate_seqs.py:
```diff
@@ -0,0 +1,76 @@
+    '''blast_results: output of blastn in outfmt == 6'''
+    alignments = []
+    results = []
+    with open(blast_results, "r") as inputfile:
```
Are `blast` results always TSV? It makes me a little nervous to re-implement a TSV parser here. Maybe this could use pandas, or python's [`csv`](https://docs.python.org/3.5/library/csv.html) instead?

nbokulich replied:
👍 pandas makes sense since the outputs are dataframes

thermokarst commented on q2_quality_control/_utilities.py:
```diff
@@ -0,0 +1,421 @@
+def _interpret_metric_selection(plot_tar, plot_tdr, plot_r_value,
+                                plot_r_squared, plot_observed_features,
+                                plot_observed_features_ratio):
+    # convert metric boolean choices to list of column names
```
👍

thermokarst commented on q2_quality_control/_utilities.py:
```diff
@@ -0,0 +1,421 @@
+            composition_regression, score_plot, mismatch_histogram)
+
+
+def _match_samples_by_index(df, df2):
```
This functionality might be slightly clearer as a [`pd.join`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.join.html) where `how='inner'`. Also, how about `df_a` and `df_b`, or `df1` and `df2`, instead of the lopsided names it is right now?

nbokulich replied:
👍 name changes

I am not sure about `pd.join`, though — I do not want to join these frames, just drop any samples that are not shared. So `pd.join` adds a extra step since I would need to do something like:
```
df_joined = pd.join([df_a, df_b], how='inner')
df_a = df_joined[[df_a.columns]]
df_b = df_joined[[df_b.columns]]
```

Is that what you had in mind?

thermokarst replied:
Oops, I misread the code the first time --- lets stick with what you have. Can you add a test where the index ids are numeric strings (e.g. `"1", "2", "3"`), just to make sure that the `.loc` behavior doesn't change out from underneath us in the future.

nbokulich replied:
👍 added `test_match_samples_by_index`

thermokarst commented on q2_quality_control/_utilities.py:
```diff
@@ -0,0 +1,421 @@
+
+
+def _find_nearest_common_lineage(feature, exp_features):
+    feature_depth = len(feature.split(';'))
```
Do we need any whitespace trimming, as well?

nbokulich replied:
That is a good point — I am not sure, but I have added it into the `str.rstrip` below just in case. Do you think we should be stripping whitespace from the list of `expected_features` as well?

thermokarst replied:
I think `expected_features` are fine, because they are being collapsed by `q2-taxa`, which performs its [own trimming](https://github.com/qiime2/q2-taxa/blob/master/q2_taxa/_util.py#L18).

thermokarst commented on q2_quality_control/_utilities.py:
```diff
@@ -0,0 +1,421 @@
+    index = join(TEMPLATES, 'index.html')
+    q2templates.render(index, output_dir, context={
+        'title': title,
+        'running-title': running_title,
```
`'running_title': running_title,`

thermokarst commented on q2_quality_control/assets/index.html:
```diff
@@ -0,0 +1,156 @@
+
+{% block content %}
+
+<div class="row">
```
Missing one more `div` between L7 and L8: `<div class="col-md-12">`. Without this bootstrap does some funny things on some resolutions.

thermokarst commented on q2_quality_control/assets/index.html:
```diff
@@ -0,0 +1,156 @@
+        <p>Download as PDF</p>
+      </a>
+    {% endif %}
+    <div class="col-lg-12">
```
Your columns (eg `col-lg-12`) should be children of semantic `row`s.

nbokulich replied:
pls elaborate. Are you saying that preceding each column I should have another `<div class="row">`?

thermokarst replied:
Can do! Functionally, [bootstrap grids](https://getbootstrap.com/docs/3.3/css/#grid) follow this pattern: a container holds rows, and rows hold columns. The `class="container"` comes from the `{% block content %}` parent template, so you don't need to do anything there, otherwise:

```jinja
{% block content %}
<h1>{{ title }}</h1>

<div class="text-center">

  <div class="row">
    <div class="col-lg-12">
      ... results ...
    </div>
  </div>

  {% if composition_regression %}
  <div class="row">
    <div class="col-lg-12">
      <h1>Linear regression between observed and expected abundances</h1>
      <a href="composition_regression.pdf">
        <img src="composition_regression.png">
        <br>
        <p>Download as PDF</p>
      </a>
    </div>
  </div>
  {% endif %}

  ... etc ...

</div>
{% endblock %}
```

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+        'expected_features': 'Expected feature compositions',
+        'observed_features': 'Observed feature compositions'},
+    parameter_descriptions={
+        'depth': ("Maximum depth of semicolon-delimited taxonomic ranks to "
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+                  "test (e.g., 1 = root, 7 = species for the greengenes "
+                  "reference sequence database)."),
+        'palette': "Color palette to utilize for plotting.",
+        'plot_tar': ("Plot taxon accuracy rate (TAR) on score plot. TAR is "
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+                     "the number of true positive features divided by the "
+                     "total number of observed features (TAR = true positives "
+                     "/ (true positives + false positives))."),
+        'plot_tdr': ("Plot taxon detection rate (TDR) on score plot. TDR is "
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+                     "the number of true positive features divided by the "
+                     "total number of expected features (TDR = true positives "
+                     "/ (true positives + false negatives))."),
+        'plot_r_value': ("Plot expected vs. observed linear regression r "
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+                     "/ (true positives + false negatives))."),
+        'plot_r_value': ("Plot expected vs. observed linear regression r "
+                         "value on score plot."),
+        'plot_r_squared': ("Plot expected vs. observed linear regression r-"
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+                         "value on score plot."),
+        'plot_r_squared': ("Plot expected vs. observed linear regression r-"
+                           "squared value on score plot."),
+        'plot_observed_features': (
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+                           "squared value on score plot."),
+        'plot_observed_features': (
+            "Plot observed features count on score plot."),
+        'plot_observed_features_ratio': (
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+            "Plot observed features count on score plot."),
+        'plot_observed_features_ratio': (
+            "Plot ratio of observed:expected features on score plot."),
+        'metadata': ('Optional sample metadata that maps observed_features '
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+        'metadata': ('Optional sample metadata that maps observed_features '
+                     'sample IDs to expected_features sample IDs.')},
+    name='Evaluate expected vs. observed taxonomic composition of samples',
+    description=(
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+    parameters={'show_alignments': Bool},
+    input_descriptions=seq_inputs_descriptions,
+    parameter_descriptions={
+        'show_alignments': ('Option to plot pairwise alignments of query '
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
+        'show_alignments': ('Option to plot pairwise alignments of query '
+                            'sequences and their top hits.')},
+    name='Compare query (observed) vs. reference (expected) sequences.',
+    description=(
```
Drop extra `( )`, please!

thermokarst commented on q2_quality_control/plugin_setup.py:
```diff
@@ -76,3 +82,87 @@
         'filtering thresholds as necessary.')
 )
+
+plugin.visualizers.register_function(
```
Can you stick to one quote mark in the following registration? We bounce back and forth between single and double quotes (I would prefer single quotes - less keystrokes).

nbokulich commented on q2_quality_control/_evaluate_seqs.py:
```diff
@@ -0,0 +1,76 @@
+def _evaluate_seqs(query_sequences, reference_sequences, show_alignments):
+    cmd = _blast(query_sequences, reference_sequences, None, 0)
+    # modify command to output seq alignment information
+    cmd[6] = ('6 qseqid sseqid pident length mismatch gapopen qstart qend '
```
👍

nbokulich commented on q2_quality_control/_evaluate_seqs.py:
```diff
@@ -0,0 +1,76 @@
+def _generate_alignments(cmd, show_alignments):
+    '''Run command line subprocess and extract hits.'''
+    with tempfile.NamedTemporaryFile() as output:
+        cmd = cmd + [output.name]
```
:man_facepalming:

nbokulich commented on q2_quality_control/_evaluate_seqs.py:
```diff
@@ -0,0 +1,76 @@
+            line = line.rstrip()
+            # qseqid sseqid pident length mismatch gapopen qstart qend
+            # sstart send evalue bitscore qlen qseq sseq
+            (qseqid, sseqid, pident, length, mismatch, gapopen, qstart, qend,
```
yes, and these fields are defined explicitly as blast output in `_evaluate_seqs`

nbokulich commented on q2_quality_control/_evaluate_seqs.py:
```diff
@@ -0,0 +1,76 @@
+    results = pd.DataFrame(results, columns=[
+            'Query id', 'Subject id', 'Percent Identity', 'Alignment Length',
+            'Mismatches', 'Gaps', 'Alignment Start (query)',
+            'Alignment end (query)', 'Alignment Start (subject)',
```
yes, 'Subject' is a blast-specific term so I am propagating that terminology. Do you think we need something more generic?

nbokulich commented on q2_quality_control/assets/index.html:
```diff
@@ -0,0 +1,156 @@
+{% extends 'base.html' %}
+
+{% block title %}q2-quality-control : {{ running_title }}{% endblock %}
```
yep, I think your fix above is the culprit. thanks!

nbokulich commented:
thanks @thermokarst ! I have pushed changes addressing all comments in your initial review.

nbokulich commented on q2_quality_control/_utilities.py:
```diff
@@ -211,7 +211,7 @@ def compute_taxon_accuracy(exp, obs):
     feature_depth = len(feature.split(';'))
     # slice off empty labels
-    feature = feature.rstrip(';_').split(';')
+    feature = feature.rstrip(';_ ').split(';')
```
sounds good — as long as this is consistent with the stripping performed by `collapse`

nbokulich commented on q2_quality_control/tests/test_quality_control.py:
```diff
@@ -278,6 +279,14 @@ def test_interpret_metric_selection_invalid(self):
+        df_b = pd.DataFrame({'a': (2., 3., 7.)}, index=['2', '3', '7'])
+        df_c = pd.DataFrame({'a': (2., 3.)}, index=['2', '3'])
+        df_a, df_b = _match_samples_by_index(df_a, df_b)
+        pdt.assert_frame_equal(df_a.dropna(), df_c)
```
`_match_sampled_by_index` actually just identifies the non-matching rows, then `_drop_nans_zeros` does the dirty work.

I have updated the comment line accordingly. I can alter the function to do this if you think that's cleaner.

nbokulich commented:
thanks @thermokarst ! Let me know what you think.
